docs/migration-guides/v2/lib.md

```diff
@@ -0,0 +1,284 @@
+
+- If you’re explicitly choosing the promises variant of the library through the subpath of the imported module — that is, if you’re writing `require('ably/promises')` — then update your code to `require('ably')`.
+- If you’re explicitly choosing the promises variant at the moment of instantiating the client — that is, if you’re writing `new Ably.Rest.Promise(…)` or `new Ably.Realtime.Promise(…)` — then update your code remove the use of the `Promise` property, i.e. write `new Ably.Realtime(…)` or `new Ably.Rest(…)`.
+
+#### Be aware of platforms that are no longer supported
+
+v2 of the library drops support for some platforms that v1 supported. Most notably, it no longer supports Internet Explorer or non-LTS versions of Node.js. For more information, see [this section of the Readme](../../../README.md#supported-platforms).
+
+<h4 id="switch-to-promise-based-generateRandomKey">Switch to using the new promise-based API of <code>Crypto.generateRandomKey()</code></h4>
+
+**Note:** This section is only relevant if you’re using the `Crypto.generateRandomKey()` method.
+
+If you’re using the `Crypto.generateRandomKey()` method, you’ll need to change how you call this method. In v1, this method required that you pass a callback. In v2, it communicates its result by returning a promise.
+
+So you need to change code that looks like this:
+
+```javascript
+Ably.Realtime.Crypto.generateRandomKey((key) => { … })
+```
+
+into code that looks like this:
+
+```javascript
+const key = await Ably.Realtime.Crypto.generateRandomKey();
+```
+
+#### Update your usage of `Rest.request()` to pass a `version` argument
+
+**Note:** This section is only relevant if you’re using the `Rest.request()` method.
+
+The signature of this method has been changed; it now requires that you pass a `version` argument to specify the major version of the REST API to use. For more information about REST API versioning, see [this section of the REST API reference](https://ably.com/docs/api/rest-api#versioning).
+
+(**TODO this wording "major version" comes from https://github.com/ably/sdk-api-reference/pull/32#pullrequestreview-1408872163. I guess that it means you should take the current sentence 'By default, all requests receive the latest version of the API, which is currently 1.1' from the linked versioning page and infer that the "major version" is 1, and so that's what I should pass to `Rest.request()`. But I thought that we've moved away from using a `major.minor` protocol version to just a single integer? See https://sdk.ably.com/builds/ably/specification/main/features/#CSV2a. Should we update those versioning docs to say that the latest version of the API is 3, and then drop the "major" in the documentation for `Rest.request()`?**)
```

Comment:
Thoughts?

Reply:
So, what the [ably.com REST API versioning documentation](https://ably.com/docs/api/rest-api#versioning) refers to as the "API version" — that is, the value of the `X-Ably-Version` request header — is the same concept as what the SDK specification calls the "protocol version" (referred to by [CSV2](https://sdk.ably.com/builds/ably/specification/main/features/#CSV2)), as evidenced by the fact that [RSC7e](https://sdk.ably.com/builds/ably/specification/main/features/#RSC7e) says that this CSV2 version is what the library should use to populate the `X-Ably-Version` header of the requests that it sends.

The Ably service accepts two different formats of protocol version:

- As the ably.com REST API versioning documentation suggests with its examples of "1.1" and "1.2", it accepts a protocol version of the format `major.minor`. This is confirmed by [CSV2a](https://sdk.ably.com/builds/ably/specification/main/features/#CSV2a), which implies that this format is accepted if `major` < 2.
- As CSV2a implies, it also accepts a single integer which is greater than or equal to 2. The maximum value permitted here will be incremented as new breaking changes are introduced to the REST or Realtime API. The implication here is that from now on, there will no longer be a concept of a "major" protocol version; there is now just "protocol version".

Now, since the `Rest.request()` mandatory version parameter was added after the introduction of protocol version 2, it accepts a single integer version. Yet, the documentation for this parameter refers to a "major" version, even though that concept no longer exists for future protocol versions. It should just refer to "the protocol version".

The problem is that in order to provide users with more information about REST API versioning, the documentation for this new `Rest.request()` version parameter links them to the ably.com REST API versioning documentation, which has not been updated recently and has a couple of problems:

- it refers to protocol versions 1.1 and 1.2, meaning that a user coming to this page from a `Rest.request()` documentation comment that mentions "the protocol version" wouldn't know how to populate this value, which is a single integer;
- it says that the latest API version is 1.1, even though [CSV2b](https://sdk.ably.com/builds/ably/specification/main/features/#CSV2b) tells us that the latest version must be at least 3

So, the ideal thing to do would be to:

- update the `Rest.request()` version parameter documentation to just refer to "the protocol version"
- update the ably.com REST API versioning documentation to say that the latest API version is 3

But, just saying "the latest version is 3" isn't very useful — what I think it really intends to say is "you are currently viewing documentation for REST API version 3". It's not very useful to know what the latest API version is without knowing how to interact with it.

So, then we'd need to update the REST API documentation to reflect the changes to the REST API made between protocol version 1.1 and 3. Off the top of my head, these are changes to the batch APIs (token revocation, batch presence, batch publish) and to the stats API (flattened, "untyped stats"). These REST API changes aren't yet documented anywhere as far as I’m aware; rather you can infer them from the corresponding changes that have been made to the SDK feature spec.

_But_, then, we'd have to figure out how to handle the problem of ably.com REST API documentation versioning. The ably.com REST API documentation isn't currently versioned, so what would happen to users who are making requests to the REST API using an `X-Ably-Version` of "1.1"? Suddenly the documentation for the API they're using would disappear.

As you can see, this simple-seeming question of how to document this new `Rest.request()` version parameter ends up requiring quite a lot of changes to the REST API documentation, so that's why I wanted to get @m-hulbert’s opinion on how to proceed.

Reply:
<del>requiring</del> potentially requiring — keen to hear other ideas

Reply:
> and I believe it would be confusing to users as well

Yeah, agreed

Reply:
Nice write up @lawrence-forooghian 👍 

The plan is to version the REST API in docs to make this easier to handle. As I understand it, the intention is to try and avoid making protocol version yet another version devs need to consider. Whilst this will work to some extent by having the docs for each language link out to the correct version of the REST API, there will still need to be some explainer text within the REST API itself as to what the version means, and by extension what this means for `rest.request()`.

Reply:
Thanks for the reply @m-hulbert. Sounds like this is stuff planned for the future. Do you have thoughts on what we should do then in this immediate context of telling users migrating to ably-js v2 how to use the version of `Rest.request()` that requires a version?

Reply:
Basically, for somebody who is now thinking "I have to specify a number here, it says; what number?!"

Reply:
I wonder if in the short term we should just say something along the lines of "set the version to `3` to use with ably-js v2". Anyone reading this guide will already be migrating to v2, so they shouldn't need to worry about the historical context of setting a version as `major.minor` rather than a single int. 

I appreciate this will be slightly contradictory to the REST API reference docs in the short term, but I think it's the simplest answer for anyone migrating.

Reply:
I think it would actually make more sense to say to use version 2 for now, since that'll provide the same behaviour as v1’s `request()`?

Reply:
i.e. when you call `request()` in v1 it sends an `X-Ably-Version: 2` header

Reply:
So, sure, that makes sense for users migrating, we say in this migration guide to specify a version of 2. As for the docstring for the method, do you want to leave it as it is (i.e. referring to "major" version and linking to the versioning docs?)

Reply:
Oh sure, if there's nothing in v3 required then I agree it makes more sense to stick with `2` as you suggest. 

For the docstring I think we can drop "major" and just refer to it as "version" since this will be the case going forwards. We can leave the link in for now too even if it'll be a little wonky until we release the versioned REST API.